File: lab1.c
/*
    This lab is all about learning to do integration using trapezoidal rule and simpson13 rule.
*/

#include <stdio.h>
#include <math.h>

#define E 2.71828
#define TRUE 1
#define FALSE 0

// definition of function1
double function1(double x){
    return 4 * pow(E, 4 * x) + 3 * pow(E, 3 * x) + 2 * pow(E, 2 * x) + pow(E, x);
}

// definition of function2
double function2(double x){
    return pow(x, 3) + 2 * x + pow(x, -1);
}

// This is a helper function used with simpson13Integration function and simply flips the isOddCoeff variable between 0 and 1. It takes in input as the pointer to that variable.
void flip(int* isOddCoeff){
    if(*isOddCoeff == 1){
        *isOddCoeff = 0;
    }
    else{
        *isOddCoeff = 1;
    }
}

// computes the integral using TRAPEZOIDAL RULE for integration. Takes in input as the function, limits and the number of segments.
double trapezoidalIntegration(double(*f)(double), double lowerLimit, double upperLimit, int numberOfSegments){
    double h = (double)(upperLimit - lowerLimit) / numberOfSegments;
    
    float value = (*f)(lowerLimit) + (*f)(upperLimit);
    for(int i = 1; i <= numberOfSegments - 1; i++){
        value += 2 * (*f)(lowerLimit + i * h);
    }
    value = value * (h / 2);
    
    return value;
}

// computes the integral using SIMPSON13 RULE for integration. Takes in input as the function, limits and the number of segments.
double simpson13Integration(double(*f)(double), double lowerLimit, double upperLimit, int numberOfSegments){
    double h = (double)(upperLimit - lowerLimit) / numberOfSegments;
    
    float value = (*f)(lowerLimit) + (*f)(upperLimit);
    int isOddCoeff = TRUE;      // this keeps track of whether we are at the odd coeff or even coeff based on whether it is TRUE OR FALSE respectively.
    int coeff;                  // changes itself to 4 or 2 based upon whether odd coeff is true or false respectively.
    for(int i = 1; i <= numberOfSegments - 1; i++, flip(&isOddCoeff)){
        coeff = isOddCoeff == TRUE ? 4 : 2;
        value += coeff * (*f)(lowerLimit + i * h);
    }
    value = value * (h / 3);
    
    return value;
}

// Driver function main
int main()
{   
    // TRAPEZOIDAL RULE for both the functions
    printf("TRAPEZOIDAL RULE / LINEAR APPROXIMATION\n");
    printf("\t-> For function1(n = 20) = %f\n", trapezoidalIntegration(function1, 1, 2, 20));
    printf("\t-> For function2(n = 20) = %f\n", trapezoidalIntegration(function2, E, 5, 20));
    
    printf("\n");
    
    // SIMPSON13 RULE for both the functions
    printf("SIMPSON13 RULE / PARABOLIC APPROXIMATION\n");
    printf("\t-> For function1(n = 20) = %f\n", simpson13Integration(function1, 1, 2, 20));
    printf("\t-> For function2(n = 20) = %f\n", simpson13Integration(function2, E, 5, 20));
    
    return 0;
}

File: lab1.py
## Deals with trapezoidal and simpson13 integration techniques

import numpy as np
import math

def f(x):
	return 4 * math.exp(4 * x) + 3 * math.exp(3 * x) + 2 * math.exp(2 * x) + math.exp(x)

def g(x):
	return x**(3) + 2 * x + x**(-1)

def trapezoidalIntegration(F, a, b, n):
	h = float(b - a) / n
	ans = F(a) + F(b)
	for i in range(1, n):
		ans += 2 * F(a + i * h)
	ans = ans * (h / 2)
	return ans
	
def simpson13Integration(F, a, b, n):
	h = float(b - a) / n
	ans = F(a) + F(b)
	isOdd = True
	for i in range(1, n):
		if(isOdd):
			coefficient = 4
		else:
			coefficient = 2
		ans += coefficient * F(a + i * h)
		isOdd = not isOdd
	ans = ans * (h / 3)
	return ans

def main():
	print 'using trapezoidal rule for 1st function with n = 10 -> ' + str(trapezoidalIntegration(f, 1, 2, 10))
	print 'using simpson13 rule for 1st function with n = 10 -> ' + str(simpson13Integration(f, 1, 2, 10))
	print 'using trapezoidal rule for 2nd function with n = 10 -> ' + str(trapezoidalIntegration(g, math.e, 5, 10))
	print 'using simpson13 rule for 2nd function with n = 10 -> ' + str(simpson13Integration(g, math.e, 5, 10))
	

if __name__ == '__main__':
	main()
